File: frontend-gym-tracker/lib/core/models/user_settings.dart
import 'enums.dart';

class UserSettings {
  const UserSettings({
    this.units = Units.metric,
    this.darkMode = true,
    this.notificationsEnabled = true,
    this.workoutRemindersEnabled = true,
    this.reminderTime = '18:00',
    this.restTimerSound = true,
    this.language = 'English',
  });

  final Units units;
  final bool darkMode;
  final bool notificationsEnabled;
  final bool workoutRemindersEnabled;

  /// 'HH:mm' 24h format.
  final String reminderTime;
  final bool restTimerSound;
  final String language;

  UserSettings copyWith({
    Units? units,
    bool? darkMode,
    bool? notificationsEnabled,
    bool? workoutRemindersEnabled,
    String? reminderTime,
    bool? restTimerSound,
    String? language,
  }) =>
      UserSettings(
        units: units ?? this.units,
        darkMode: darkMode ?? this.darkMode,
        notificationsEnabled: notificationsEnabled ?? this.notificationsEnabled,
        workoutRemindersEnabled:
            workoutRemindersEnabled ?? this.workoutRemindersEnabled,
        reminderTime: reminderTime ?? this.reminderTime,
        restTimerSound: restTimerSound ?? this.restTimerSound,
        language: language ?? this.language,
      );

  Map<String, dynamic> toJson() => {
        'units': units.name,
        'darkMode': darkMode,
        'notificationsEnabled': notificationsEnabled,
        'workoutRemindersEnabled': workoutRemindersEnabled,
        'reminderTime': reminderTime,
        'restTimerSound': restTimerSound,
        'language': language,
      };

  factory UserSettings.fromJson(Map<String, dynamic> json) => UserSettings(
        units: Units.fromName(json['units'] as String?),
        darkMode: json['darkMode'] as bool? ?? true,
        notificationsEnabled: json['notificationsEnabled'] as bool? ?? true,
        workoutRemindersEnabled:
            json['workoutRemindersEnabled'] as bool? ?? true,
        reminderTime: json['reminderTime'] as String? ?? '18:00',
        restTimerSound: json['restTimerSound'] as bool? ?? true,
        language: json['language'] as String? ?? 'English',
      );
}

File: backend-gym-tracker/src/main/java/com/patriche/gymtracker/settings/UserSettings.java
package com.patriche.gymtracker.settings;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.Instant;
import java.util.UUID;

/**
 * One row per user. Defaults match UserSettings in the Flutter app, so a user who has
 * never opened the settings screen syncs the same values the app would have shown.
 *
 * <p>Not a SyncableEntity: settings are a single row that is always overwritten whole,
 * so there is nothing to tombstone.
 */
@Entity
@Table(name = "user_settings")
public class UserSettings {

    @Id
    @Column(name = "user_id")
    private UUID userId;

    @Column(nullable = false)
    private String units = "metric";

    @Column(name = "dark_mode", nullable = false)
    private boolean darkMode = true;

    @Column(name = "notifications_enabled", nullable = false)
    private boolean notificationsEnabled = true;

    @Column(name = "workout_reminders_enabled", nullable = false)
    private boolean workoutRemindersEnabled = true;

    /** 'HH:mm' 24h, stored as text to match the app exactly. */
    @Column(name = "reminder_time", nullable = false)
    private String reminderTime = "18:00";

    @Column(name = "rest_timer_sound", nullable = false)
    private boolean restTimerSound = true;

    @Column(nullable = false)
    private String language = "English";

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt = Instant.EPOCH;

    protected UserSettings() {}

    public UserSettings(UUID userId, Instant now) {
        this.userId = userId;
        this.updatedAt = now;
    }

    public UUID getUserId() { return userId; }
    public String getUnits() { return units; }
    public boolean isDarkMode() { return darkMode; }
    public boolean isNotificationsEnabled() { return notificationsEnabled; }
    public boolean isWorkoutRemindersEnabled() { return workoutRemindersEnabled; }
    public String getReminderTime() { return reminderTime; }
    public boolean isRestTimerSound() { return restTimerSound; }
    public String getLanguage() { return language; }
    public Instant getUpdatedAt() { return updatedAt; }

    public void setUnits(String v) { this.units = v; }
    public void setDarkMode(boolean v) { this.darkMode = v; }
    public void setNotificationsEnabled(boolean v) { this.notificationsEnabled = v; }
    public void setWorkoutRemindersEnabled(boolean v) { this.workoutRemindersEnabled = v; }
    public void setReminderTime(String v) { this.reminderTime = v; }
    public void setRestTimerSound(boolean v) { this.restTimerSound = v; }
    public void setLanguage(String v) { this.language = v; }
    public void touch(Instant now) { this.updatedAt = now; }
}
